File: ocaml/simpleaudio_stubs.ml
external start_audio: unit -> unit = "caml_start_audio"
external oscilloscope: unit -> unit = "caml_oscilloscope"

external kill_audio: unit -> unit = "caml_kill_audio" 


external add_node: unit -> unit = "caml_add_node" 

type node_ptr

external osc_sin: float -> node_ptr = "caml_sin"
module StringMap = Map.Make(String)
type param_map = int StringMap.t

let make_param_map keys =
  let rec aux counter acc = function
    | [] -> acc
    | hd :: tl ->
        let acc' = StringMap.add hd counter acc in
        aux (counter + 1) acc' tl
  in
  aux 0 StringMap.empty keys
type node = { param_map: param_map; node_ptr: node_ptr }

type node_param = 
  | Float of float
  | Node of node

let set_param node key value =
  let updated_param_map = StringMap.add key value node.param_map in
  { node with param_map = updated_param_map }

let sin x =
  let node_ptr = osc_sin x in
  let param_map = make_param_map ["freq"] in
  {
    param_map = param_map;
    node_ptr = node_ptr; 
  }

let node_param_idx node param_name =
  StringMap.find param_name node.param_map


external set_sig_float: node_ptr -> int -> float -> unit = "caml_set_sig_float"
external set_sig_node: node_ptr -> int -> node_ptr -> unit = "caml_set_sig_node"
external caml_set_mul_float: node_ptr -> float -> unit = "caml_set_mul" 
external caml_set_mul_node: node_ptr -> node_ptr -> unit = "caml_set_mul_node" 
external caml_set_add_float: node_ptr -> float -> unit = "caml_set_add" 
external caml_set_add_node: node_ptr -> node_ptr -> unit = "caml_set_add_node" 
let set_float node param_name f = 
  (match param_name with
  | "mul" -> caml_set_mul_float node.node_ptr f
  | "add" -> caml_set_add_float node.node_ptr f
  | _ -> 
    let param_idx = node_param_idx node param_name in 
    set_sig_float node.node_ptr param_idx f);
  node



let set_node node param_name n = 
  (match param_name with
  | "mul" -> caml_set_mul_node node.node_ptr n.node_ptr
  | "add" -> caml_set_add_node node.node_ptr n.node_ptr
  | _ -> 
    let param_idx = node_param_idx node param_name in 
    set_sig_node node.node_ptr param_idx n.node_ptr);
  node


let set node param_name value = 
  match value with
  | Float f -> set_float node param_name f
  | Node n -> set_node node param_name n
  



external set_freq: node_ptr -> float -> unit = "caml_set_freq"
external osc_sq: float -> node_ptr = "caml_sq"

let sq x =
  let node_ptr = osc_sq x in

  let param_map = make_param_map ["freq"; "pw"] in
  {
    param_map = param_map;
    node_ptr = node_ptr; 
  }

external osc_sq_detune: float -> node_ptr = "caml_sq_detune"

let sq_detune x =
  let node_ptr = osc_sq_detune x in
  let param_map = (
    StringMap.empty
    |> StringMap.add "freq" 0 
    |> StringMap.add "pw" 1
    |> StringMap.add "amp" 2
  ) in
  {
    param_map = param_map;
    node_ptr = node_ptr; 
  }

external osc_impulse: float -> node_ptr = "caml_impulse"

let impulse x =
  let node_ptr = osc_impulse x in

  let param_map = make_param_map ["freq"] in
  {
    param_map = param_map;
    node_ptr = node_ptr; 
  }

external osc_saw: float -> node_ptr = "caml_poly_saw"

let saw x =
  let node_ptr = osc_saw x in
  let param_map = make_param_map ["freq"] in
  {
    param_map = param_map;
    node_ptr = node_ptr; 
  }

external osc_hoover: float -> node_ptr = "caml_hoover"
external osc_pulse: float -> float -> node_ptr = "caml_pulse"

let pulse f pw =
  let node_ptr = osc_pulse f pw in

  let param_map = make_param_map ["freq"; "pw"] in

  {
    param_map = param_map;
    node_ptr = node_ptr; 
  }


external fx_delay: float -> float -> node_ptr = "caml_simple_delay"
let delay t f =
  let node_ptr = fx_delay t f in 

  let param_map = make_param_map ["time"; "fb"] in
  {
    param_map = param_map;
    node_ptr = node_ptr; 
  }

external caml_chain_nodes: node_ptr -> node_ptr -> int -> node_ptr = "caml_chain_nodes" 

external stop: node_ptr -> unit = "caml_kill_node"
external dump_nodes: unit -> unit = "caml_dump_nodes"

let make_node node_ptr param_map =
  {param_map = param_map; node_ptr = node_ptr;}

let chain_nodes a b ?(p=0) =
  let node_ptr = caml_chain_nodes a.node_ptr b.node_ptr p in
  (
    make_node node_ptr a.param_map,
    make_node b.node_ptr b.param_map
  )
;;

external play_node: node_ptr -> node_ptr = "caml_play_node"
let play node = 
  play_node node.node_ptr

let chain a b p =
  let node_ptr = caml_chain_nodes a.node_ptr b.node_ptr (node_param_idx b p) in
  ({
    param_map = a.param_map;
    node_ptr = node_ptr; 
  }, 
  {
    param_map = b.param_map;
    node_ptr = b.node_ptr; 
  })
;;


(* Define the chaining operator => *)
let (=>) a b = chain_nodes a b ~p:0
let (=>.) a p b = chain_nodes a b ~p:(node_param_idx b p)

(* external extern_set_list_float: node_ptr -> unit = "caml_set_list_float" *)

(* let set_list node mapping_list = *)
(*   let f el = *)
(*     let (p, v) = el in *)
(*     (node_param_idx node p, v); *)
(*  *)
(*   extern_set_list_float node.node_ptr (List.map f mapping_list) *)

File: ocaml/simpleaudio_stubs.c
#include "../src/audio/delay.h"
#include "../src/audio/osc.h"
#include "../src/audio/out.h"
#include "../src/ctx.h"
#include "../src/log.h"
#include "../src/node.h"
#include "../src/oscilloscope.h"
#include "../src/start_audio.h"
#include <caml/alloc.h>
#include <caml/memory.h>
#include <caml/mlvalues.h>
#include <stdio.h>

#include <pthread.h>

static value Val_node(Node *p) { return caml_copy_nativeint((intnat)p); }
static Node *Node_val(value v) { return (Node *)Nativeint_val(v); }
CAMLprim value caml_start_audio() {
  int audio_status = setup_audio();
  return Val_unit;
}

CAMLprim value caml_kill_audio() {
  int audio_status = stop_audio();
  return Val_unit;
}

CAMLprim value caml_chain_nodes(value source_ptr, value dest_ptr, value idx) {
  Node *source = Node_val(source_ptr);
  Node *dest = Node_val(dest_ptr);
  chain_nodes(source, dest, Int_val(idx));
  return Val_node(source);
}

static Node *wrap_chain(Node *head, Node *tail) {
  Node *container = ALLOC_NODE(container_node_data, "Container");
  container->_sub = head;
  container->_sub_tail = tail; // point to the last node in the chain
                               // that's not an add_out or replace_out -
                               // could be the same node as head
  return container;
}

CAMLprim value caml_sin(value freq) {
  Node *osc = sin_node(Double_val(freq), NULL);
  /* Node *container = container_node(osc); */
  /* ctx_add_after_tail(container); */
  return Val_node(osc);
}

CAMLprim value caml_sq(value freq) {
  Node *osc = sq_node(Double_val(freq), NULL);
  /* Node *container = container_node(osc); */
  /* ctx_add_after_tail(container); */
  return Val_node(osc);
}

CAMLprim value caml_sq_detune(value freq) {
  Node *osc = sq_detune_node(Double_val(freq), NULL);
  /* Node *container = container_node(osc); */
  /* ctx_add_after_tail(container); */
  return Val_node(osc);
}

CAMLprim value caml_impulse(value freq) {

  Node *osc = impulse_node(Double_val(freq), NULL);
  /* Node *container = container_node(osc); */
  /* ctx_add_after_tail(container); */
  return Val_node(osc);
}

CAMLprim value caml_poly_saw(value freq) {

  Node *osc = poly_saw_node(Double_val(freq), NULL);

  /* Node *container = container_node(osc); */
  /* ctx_add_after_tail(container); */

  return Val_node(osc);
}

CAMLprim value caml_pulse(value freq, value pw) {

  Node *osc = pulse_node(Double_val(freq), Double_val(pw), NULL);
  /* Node *container = container_node(osc); */
  /* ctx_add_after_tail(container); */

  return Val_node(osc);
}

CAMLprim value caml_simple_delay(value time, value fb) {
  Node *osc = simple_delay_node(Double_val(time), Double_val(fb), 1.0, NULL);
  return Val_node(osc);
}

CAMLprim value caml_oscilloscope() {
  /* pthread_t oscil_thread; */
  /* pthread_create(&oscil_thread, NULL, oscilloscope, NULL); */
  /* pthread_detach(oscil_thread); */
  oscilloscope();
  return Val_unit;
}

CAMLprim value caml_set_freq(value ml_ptr, value d) {
  Node *node;
  node = Node_val(ml_ptr);
  Signal freq = IN(node, SIN_SIG_FREQ);
  set_signal(freq, Double_val(d));
  return Val_unit;
}

CAMLprim value caml_set_sig_float(value ml_ptr, value sig, value d) {
  Node *node = Node_val(ml_ptr);
  int sig_idx = Int_val(sig);

  if (sig_idx >= node->num_ins) {
    printf("node has no input %d\n", sig_idx);
    return Val_unit;
  }
  Signal signal = IN(node, sig_idx);
  set_signal(signal, Double_val(d));
  return Val_unit;
}

CAMLprim value caml_set_sig_node(value ml_ptr, value sig_num, value d) {
  Node *node = Node_val(ml_ptr);
  int sig_idx = Int_val(sig_num);
  Node *src = Node_val(d);

  if (sig_idx >= node->num_ins) {
    printf("node has no input %d\n", sig_idx);
    return Val_unit;
  }

  Signal input = IN(node, sig_idx);

  input.data = src->out.data;
  input.size = src->out.size;

  return Val_unit;
}

CAMLprim value caml_kill_node(value ml_ptr) {
  Node *node = Node_val(ml_ptr);
  node->killed = true;
  return Val_unit;
}

typedef struct {
  int x;
  char *y;
} obj;

typedef struct _obj_st {
  double d;
  int i;
  char c;
} obj_st;

static obj_st *Objst_val(value v) { return (obj_st *)Nativeint_val(v); }
static value Val_objst(obj_st *p) { return caml_copy_nativeint((intnat)p); }

CAMLprim value wrapping_ptr_ml2c(value d, value i, value c) {
  obj_st *my_obj;
  my_obj = malloc(sizeof(obj_st));
  my_obj->d = Double_val(d);
  my_obj->i = Int_val(i);
  my_obj->c = Int_val(c);
  return Val_objst(my_obj);
}
CAMLprim value set_d(value ml_ptr, value d) {
  obj_st *my_obj;
  my_obj = Objst_val(ml_ptr);
  my_obj->d = Double_val(d);
  printf("set d to %f\n", my_obj->d);
  return Val_unit;
}

CAMLprim value dump_ptr(value ml_ptr) {
  obj_st *my_obj;
  my_obj = Objst_val(ml_ptr);
  printf(" d: %g\n i: %d\n c: %c\n", my_obj->d, my_obj->i, my_obj->c);
  return Val_unit;
}

CAMLprim value free_ptr(value ml_ptr) {
  obj_st *my_obj;
  my_obj = Objst_val(ml_ptr);
  free(my_obj);
  return Val_unit;
}

#define INDENT(indent) write_log("%.*s", indent, "  ")
static void dump_nodes(Node *head, int indent) {
  if (!head) {
    return;
  }
  INDENT(indent);
  write_log("%s num inputs: %d", head->name, ((signals *)head->data)->num_ins);

  if (head->_sub) {
    write_log("\n");
    dump_nodes(head->_sub, indent + 1);
  }
  if (head->next) {
    write_log("\n");
    dump_nodes(head->next, indent);
  }
  write_log("\n");
  return;
}

CAMLprim value caml_dump_nodes() {
  dump_nodes(ctx.head, 0);
  fflush(log_stream);
  return Val_unit;
}

CAMLprim value caml_set_add(value ml_ptr, value sig) {
  Node *node = Node_val(ml_ptr);
  double val = Double_val(sig);

  set_signal(node->add, val);
  return Val_unit;
}

CAMLprim value caml_set_mul(value ml_ptr, value sig) {
  Node *node = Node_val(ml_ptr);
  double val = Double_val(sig);

  set_signal(node->mul, val);
  return Val_unit;
}

CAMLprim value caml_set_add_node(value ml_ptr, value sig) {
  Node *node = Node_val(ml_ptr);
  Node *src = Node_val(sig);

  node->add.data = src->out.data;
  node->add.size = src->out.size;

  return Val_unit
}

CAMLprim value caml_set_mul_node(value ml_ptr, value sig) {
  Node *node = Node_val(ml_ptr);
  Node *val = Node_val(sig);
  node->mul.data = val->out.data;
  node->mul.size = val->out.size;

  return Val_unit;
}

CAMLprim value caml_play_node(value node_ptr) {
  Node *node = Node_val(node_ptr);

  Node *container = container_node(node);
  ctx_add_after_tail(container);

  return Val_node(container);
}

CAMLprim value caml_set_list_float(value ml_ptr, value l) {
  Node *node = Node_val(ml_ptr);
  /* int sig_idx = Int_val(sig); */
  /*  */
  /* if (sig_idx >= node->num_ins) { */
  /*   printf("node has no input %d\n", sig_idx); */
  /*   return Val_unit; */
  /* } */
  /* Signal signal = IN(node, sig_idx); */
  /* set_signal(signal, Double_val(d)); */
  return Val_unit;
}
